worlds/civ_6/__init__.py

```diff
@@ -0,0 +1,325 @@
+    )
+)
+
+
+class CivVIWeb(WebWorld):
+    tutorials = [
+        Tutorial(
+            "Multiworld Setup Guide",
+            "A guide to setting up Civilization VI for MultiWorld.",
+            "English",
+            "setup_en.md",
+            "setup/en",
+            ["hesto2"],
+        )
+    ]
+    theme = "ocean"
+
+
+class CivVIWorld(World):
+    """
+    Civilization VI is a turn-based strategy video game in which one or more players compete alongside computer-controlled opponents to grow their individual civilization from a small tribe to control the entire planet across several periods of development.
+    """
+
+    game = "Civilization VI"
+    topology_present = False
+    options_dataclass = CivVIOptions
+    options: CivVIOptions  # type: ignore
+
+    web = CivVIWeb()
+
+    item_name_to_id = {item.name: item.code for item in generate_item_table().values()}
+    location_name_to_id = {
+        location.name: location.code
+        for location in generate_flat_location_table().values()
+    }
+
+    item_table: Dict[str, CivVIItemData] = {}
+    location_by_era: Dict[str, Dict[str, CivVILocationData]]
+    required_client_version = (0, 4, 5)
+    location_table: Dict[str, CivVILocationData]
+    era_required_non_progressive_items: Dict[EraType, List[str]] = {}
+    era_required_progressive_items_counts: Dict[EraType, Dict[str, int]] = {}
+    era_required_progressive_era_counts: Dict[EraType, int] = {}
+    item_by_civ_name: Dict[str, str]
+
+    def __init__(self, multiworld: MultiWorld, player: int):
+        super().__init__(multiworld, player)
+        self.location_by_era = generate_era_location_table()
+
+        self.location_table: Dict[str, CivVILocationData] = {}
+        self.item_table = generate_item_table()
+
+        for locations in self.location_by_era.values():
+            for location in locations.values():
+                self.location_table[location.name] = location
+
+    def generate_early(self) -> None:
+        flat_progressive_items = get_flat_progressive_districts()
+        progressive_districts = get_progressive_districts_data()
+
+        self.item_by_civ_name = {
+            item.civ_name: get_item_by_civ_name(item.civ_name, self.item_table).name
+            for item in self.item_table.values()
+            if item.civ_name
+        }
+
+        previous_era_counts = None
+        eras_list = [e.value for e in EraType]
+        for era in EraType:
+            # Initialize era_required_progressive_era_counts
+            era_index = eras_list.index(era.value)
+            self.era_required_progressive_era_counts[era] = (
+                0
+                if era in {EraType.ERA_FUTURE, EraType.ERA_INFORMATION}
+                else era_index + 1
+            )
+
+            # Initialize era_required_progressive_items_counts
+            self.era_required_progressive_items_counts[era] = {
+                format_item_name(item): 0 for item in progressive_districts.keys()
+            }

```

Comment:
These dicts ends up having a lot of 0s.

This will slow down the call to `state.has_item_counts` unnecessarily, which now has to iterate all the 0s for no reason.

It would be better to use a `.setdefault` or `defaultdict` pattern here.

Reply:
(There was a thing previously where I suggested having entries with `0`, but that was because you had a condition that was explicitly checking for them. In this case it's just bad)

Reply:
Ah nice, that's cool you can do math on a defaultdict and it treats it like a zero 🤯